###Construya una  función cuyas entradas sean:
1. Una lista con los posibles resultados de lanzar una moneda. Por ejemplo ['C','S']
2. Un número de lanzamientos. 

La función debe lanzar la moneda N veces, y al final la salida debería mostrar la fracción de caras y la fracción de sellos. 

¿Qué pasa si lanza la moneda 20, 50, 100, 500 o 1000 veces?

In [6]:
import random

def imprimir(x, y):
  posibles = ['C', 'S']
  valores = []
  for i in range(0, y):
    valores.append(random.choice(posibles))
  print(valores)
  return valores.count(x)


while True:
  try:
    posibles = ['C', 'S']
    x = input("Digite la opcion : ")
    if x in (posibles):
      y = int(input('Cantidad de veces : '))
      print(imprimir(x, y))
    else:
      print('valor incorrecto')
  except:
    break

Repita el ejercicio anterior, pero esta vez, para cada uno de los valores de N, lance la moneda dos veces, y agrupe los resultados en una lista tipo ['CS','CC',...].
Al final de la rutina diga la fracción de cada resultado

In [7]:
import random

def imprimir2(y):
  posibles = ['C', 'S']
  valores = []
  for i in range(0, y):
    valores.append(random.choice(posibles) + random.choice(posibles))
  print(valores)
  print("CC -> ",(valores.count('CC') / y))
  print("CS -> ",(valores.count('CS') / y))
  print("SC -> ",(valores.count('SC') / y))
  print("SS -> ",(valores.count('SS') / y))

while True:
  try:
    y = int(input('Cantidad de veces : '))
    imprimir2(y)
  except:
    break


Imagine que tiene una cuadrícula de 3x3 donde cada cuadrado está marcado de la forma:

0 1 2

3 4 5

6 7 8

En cada cuadro, es posible moverse al azar a la izquierda, derecha, arriba o abajo. La regla es que si está en el extremo izquierdo, no puede moverse a la izquierda; si está arriba, no puede moverse hacia arriba, etc. 

Por ejemplo, si está en 4, los posibles movimientos serían [5,1,3,7]; si está en 1, los psoibles movimientos serían [2,1,0,4], etc.

Construya una rutina que muestre: 
1. Una lista de listas (o un diccionario) que muestre los posibles movimientos desde cada baldosa en el orden (derecha, arriba, izquierda, abajo). 
2. Elija un número N de movimientos.
3. Párese en una baldosa al azar. 
4. A partir de esa baldosa, muévase al azar en una dirección dada. 
5. Repita eso N veces.

Al final, cuente cuántas veces pasó por cada baldosa.

In [10]:
movimientos=['Arriba','Abajo','Derecha','Izquierda']
matris = []
posicion = 0
maximo = 0
asterisco = 0
asteriscoX = 0
asteriscoY = 0

def crearMatrix(x,y):
  n = 0
  for i in range(x):
    filas = []
    for j in range(y):
      filas.append(n)
      n += 1
    matris.append((filas))  
  return n

def imprimirmatris():
  print("")
  for i in (matris):
    print(i)
  print("")

def pedir():
  global posicion
  print("maximo -->  ", maximo)
  numero = maximo + 1  
  while numero > maximo:
    numero = int(input("posicion inicial : "))
    if numero > maximo:
      print("¡intente de nuevo!")
  print("")
  posicion = numero

def ponerasterisco(x,y):
  global asteriscoY
  global asteriscoX
  for i in range(x):    
    for j in range(y):      
      if posicion == matris[i][j]:        
        matris[i][j] = "*"
        asteriscoX = int(i) 
        asteriscoY = int(j)     
        break
    if posicion == matris[i][j]:
        break

def quitarasterisco(x,y):
  for i in range(x):    
    for j in range(y):
      if "*" == matris[i][j]:
        matris[i][j] = posicion
        break
    if "*" == matris[i][j]:
        break
      
def moverse(mover,x,y):
  global posicion
  global asteriscoY
  global asteriscoX
  if mover == "Derecha":
    if (asteriscoY+1) >= y:
      print("No se puede correr a la Derecha")
    else:
      posicion += 1      
      print("Se movio a la Derecha")
  elif mover == "Izquierda":
    if (asteriscoY - 1) < 0:
      print("No se puede correr a la Izquierda")
    else:
      posicion -= 1
      print("Se movio a la Izquierda")
  elif mover == "Arriba":
    if (asteriscoX-1) < 0:
      print("No se puede correr hacia Arriba")
    else:
      posicion -= y 
      print("Se movio hacia Arriba")
  elif mover == "Abajo":
    if (asteriscoX+1) >= x:
      print("No se puede correr hacia Abajo")
    else:
      posicion += y   
      print("Se movio hacia Abajo")
  
while True:
  try:    
    x = int(input("Digite la cantidad de filas : "))
    y = int(input("Digite la cantidad de columnas : "))    
    maximo = (crearMatrix(x,y) - 1)
    imprimirmatris()    
    pedir()
    ponerasterisco(x,y)
    imprimirmatris()
    quitarasterisco(x,y)
    while True:
      try:        
        print("Selecciona un movimiento")
        print("Derecha")
        print("Izquierda")
        print("Arriba")
        print("Abajo")
        print("")
        m = input("")
        if m in (movimientos):
          moverse(m,x,y)
          ponerasterisco(x,y)
          imprimirmatris()
          quitarasterisco(x,y)
        else:
          print("¡seleccion incorrecta !")
      except:
         break
  except:
    break


Digite la cantidad de filas : 6
Digite la cantidad de columnas : 6

[0, 1, 2, 3, 4, 5]
[6, 7, 8, 9, 10, 11]
[12, 13, 14, 15, 16, 17]
[18, 19, 20, 21, 22, 23]
[24, 25, 26, 27, 28, 29]
[30, 31, 32, 33, 34, 35]

maximo -->   35
posicion inicial : 15


[0, 1, 2, 3, 4, 5]
[6, 7, 8, 9, 10, 11]
[12, 13, 14, '*', 16, 17]
[18, 19, 20, 21, 22, 23]
[24, 25, 26, 27, 28, 29]
[30, 31, 32, 33, 34, 35]

Selecciona un movimiento
Derecha
Izquierda
Arriba
Abajo

Derecha
Se movio a la Derecha

[0, 1, 2, 3, 4, 5]
[6, 7, 8, 9, 10, 11]
[12, 13, 14, 15, '*', 17]
[18, 19, 20, 21, 22, 23]
[24, 25, 26, 27, 28, 29]
[30, 31, 32, 33, 34, 35]

Selecciona un movimiento
Derecha
Izquierda
Arriba
Abajo

Izquierda
Se movio a la Izquierda

[0, 1, 2, 3, 4, 5]
[6, 7, 8, 9, 10, 11]
[12, 13, 14, '*', 16, 17]
[18, 19, 20, 21, 22, 23]
[24, 25, 26, 27, 28, 29]
[30, 31, 32, 33, 34, 35]

Selecciona un movimiento
Derecha
Izquierda
Arriba
Abajo

Arriba
Se movio hacia Arriba

[0, 1, 2, 3, 4, 5]
[6, 7, 8, '*', 10, 11]
[12, 13, 14, 15,

Palabras palindromas

"in girum imus nocte et consumimur igni", 'Atar a la rata','¿Acaso hubo búhos acá?','¿Son robos o sobornos?','A ti no, bonita.'

In [12]:
import string

def normalize(s):
    replacements = (
        ("á", "a"),
        ("é", "e"),
        ("í", "i"),
        ("ó", "o"),
        ("ú", "u"),
    )
    for a, b in replacements:
        s = s.replace(a, b).replace(a.upper(), b.upper())
    return s
  
def f(x):
  x = normalize(x)
  quitar = string.punctuation + "¡" + "¿" 
  texto = ""
  for i in (x):
    if i != " ": 
      if not i in(quitar):
        texto = texto + i  
  return (texto.upper() == texto[::-1].upper())

while True:
  try:
    x = (input("Digite la palabra "))   
    print(f(x))
  except:
    break


**Funcion Binomial 1**

En esta función, n, p y k son los tres argumentos de la función, que representan el número de ensayos, la probabilidad de éxito en cada ensayo y el número de éxitos, respectivamente.

La función utiliza la función math.comb() para calcular el número de combinaciones de n ensayos y k éxitos. A continuación, la función utiliza la fórmula de la distribución binomial para calcular la probabilidad de k éxitos dados n ensayos y una probabilidad de éxito p. La fórmula incluye el cálculo del término binomial, la probabilidad de éxito y la probabilidad de fracaso.

In [15]:
def binomial(n, p, k):
    import math
    return math.comb(n, k) * p**k * (1-p)**(n-k)

n = 10
p = 0.5
k = 5
resultado = binomial(n, p, k)
print(resultado) # output: 0.24609375


0.24609375


En este ejemplo, se calcula la probabilidad de obtener 5 caras en 10 lanzamientos de una moneda equilibrada (es decir, una probabilidad de éxito de 0.5). El resultado es aproximadamente 0.25.

**Funcion Binomial 2**

En esta función, n, p y k son los tres argumentos de la función, que representan el número de ensayos, la probabilidad de éxito en cada ensayo y el número de éxitos, respectivamente.

La función factorial() es una implementación propia de la función factorial. Esta función utiliza la recursividad para calcular el factorial de un número dado.

A continuación, la función utiliza la fórmula de la distribución binomial para calcular la probabilidad de k éxitos dados n ensayos y una probabilidad de éxito p. La fórmula incluye el cálculo del término binomial, la probabilidad de éxito y la probabilidad de fracaso.

In [19]:
def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n-1)

def binomial(n, p, k):
    comb = factorial(n) // (factorial(k) * factorial(n-k))
    return comb * p**k * (1-p)**(n-k)

n = 10
p = 0.5
k = 5
resultado = binomial(n, p, k)
print(resultado) # output: 0.24609375


0.24609375


En este ejemplo, se calcula la probabilidad de obtener 5 caras en 10 lanzamientos de una moneda equilibrada (es decir, una probabilidad de éxito de 0.5). El resultado es aproximadamente 0.25.

**Funcion Poisson,**

En esta función, lam y k son los dos argumentos de la función, que representan el parámetro de la distribución Poisson y el número de eventos, respectivamente.

La función utiliza la constante matemática e (aproximadamente igual a 2.718) para calcular la función exponencial. A continuación, la función utiliza la fórmula de la distribución Poisson para calcular la probabilidad de que ocurran k eventos dados un parámetro de lam. La fórmula incluye el cálculo del factorial de k.

In [ ]:
def poisson(lam, k):
    import math
    e = math.exp(1)
    return (lam ** k / math.factorial(k)) * e ** (-lam)

lam = 2
k = 3
resultado = poisson(lam, k)
print(resultado) 

En este ejemplo, se calcula la probabilidad de que ocurran 3 eventos en un intervalo de tiempo dado si el número medio de eventos por intervalo de tiempo es de 2. El resultado es aproximadamente 0.18.

**Funcion Gaussiana,**

En esta función, x, mu y sigma son los tres argumentos de la función, que representan el valor de la variable, la media y la desviación estándar de la distribución Gaussiana, respectivamente.

La función utiliza la constante matemática pi (aproximadamente igual a 3.14159) para calcular la función de densidad de probabilidad Gaussiana. A continuación, la función utiliza la fórmula de la distribución Gaussiana para calcular la probabilidad de que x sea igual a un valor dado dado una media mu y una desviación estándar sigma. La fórmula incluye el cálculo de la exponencial y la raíz cuadrada.

In [9]:
def gaussiana(x, mu, sigma):
    import math
    return math.exp(-(x - mu)**2 / (2 * sigma**2)) / (sigma * math.sqrt(2 * math.pi))

x = 1
mu = 0
sigma = 1
resultado = gaussiana(x, mu, sigma)
print(resultado) 

0.24197072451914337


En este ejemplo, se calcula la probabilidad de que x sea igual a 1 dada una media de 0 y una desviación estándar de 1. El resultado es aproximadamente 0.24.